### ISSUE TITLE
[BERT/PyTorch] Incorrect next_sentence_labels when length of corpora = 1

### ISSUE THREAD

**Issue by adityaiitb** [NVIDIA/DeepLearningExamples#421]
### Create an input file a.txt. Omitting full stops to keep it simple. Also length of corpora is 1.
```
hello
tom sawyer
how are you
```

### Run it through the `create_pretraining_data.py` to create an HDF5 file.
```sh
$ a=~/DeepLearningExamples/PyTorch/LanguageModeling/BERT
$ python $a/create_pretraining_data.py --vocab_file $a/vocab/vocab --input_file a.txt --output_file b.h5 --max_seq_length 40
```

### Dump HDF5 file to see raw data
```sh
$ h5dump b.h5
```

In the dump, there are multiple instances where "valid" next sentences are marked "random". For example:

A = "hello tom sawyer", B = "how are you" is marked random (sentence id 5)
A  = "tom sawyer", B = "how are you" is marked random (sentence id 15)

This stems from the fact that after the for loop on line 247, `random_document_index == document_index` when the number of documents = 1 (as in this toy example). Therefore, the segment B is picked from the same document as segment A i.e. document 0. Since this was a toy example with only 3 sentences, even after randomization, valid sequences got constructed and were marked as invalid. The problematic sentence ids might differ on your machine.

**Comment by adityaiitb** [comment]
The code for TF is also almost the same and so would have the same issue, though I did not verify it.

https://github.com/NVIDIA/DeepLearningExamples/blob/master/TensorFlow/LanguageModeling/BERT/utils/create_pretraining_data.py

**Comment by nvcforster** [comment]
Hi Aditya,

Thanks for reporting this problem. We will take a look into it. I do want to point out that the file create_pretraining_data.py originated in the Google Research BERT repository, which is directly used in our BERT-TF example and our BERT-PyT example just tacks on the pieces for saving to hdf5 rather than tfrecords. If confirmed, there is a good chance the issue exists in all repositories.

https://github.com/google-research/bert/blob/master/create_pretraining_data.py#L279

Same goes for MLPerf/training, https://github.com/mlperf/training/blob/master/language_model/tensorflow/bert/create_pretraining_data.py#L262

**Comment by sharathts** [comment]
You technically cannot generate data for NextSentencePrediction with just one document. You need to have multiple documents to be able to pick a random sentence from a random document.

I believe you are facing the same, trying to pick a random sentence but from only one document, so both your segments corresponds to the same document even though the code decides to pick a random sentence.

**Comment by adityaiitb** [comment]
I don't think that is true. There is nothing fundamental or technical about picking the next sentence strictly from other documents. One can pick two random sentences (non-contiguous) from a single document. There is no requirement that BERT or any language model has to be trained on multiple documents. One could train it on a single huge document for instance.

What the code does (and what you described) is that it adopts one possible (and reasonable) approach of picking a random sentence, by saying "lets look at another random document". This *generally* results in a random next sentence. However, even this process is not 100% accurate because its possible that the sentence chosen from a random document might be the same as the next sentence in the current document.

Here is a very simple example
```vim
# document 1
hello
tom sawyer
how are you

# document 2
juliet
i have not seen you in a long time
how are you
```
I can pick `tom sawyer` from the first document and `how are you` from the second document and the code will mark them as random next sentence, but in reality they are sequential.

**Comment by sharathts** [comment]
Yes, agreed. The codebase has been fixed to generate correct NSP labels even in with a single doc. https://github.com/NVIDIA/DeepLearningExamples/pull/426

Also, the current setup is to pick the next sentence from another document, not the same document. 

Thanks for bringing this up!
